File: ports/h4_core.c
// H4 planogram engine core — pure integer, no float. C reference for the
// 3-language benchmark. Bit-exact with the Python/Rust implementations.
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <time.h>

#define BIAS 512u
// H4 encode: (x,y,w,h) u8 -> (X,Y,Z,W) u16, 8 integer add/sub, unsigned
static inline void h4_encode(const uint8_t b[4], uint16_t e[4]){
    uint32_t x=b[0],y=b[1],w=b[2],h=b[3];
    e[0]=(uint16_t)(x+y+w+h);
    e[1]=(uint16_t)((x+w+BIAS)-(y+h));
    e[2]=(uint16_t)((x+y+BIAS)-(w+h));
    e[3]=(uint16_t)((x+h+BIAS)-(y+w));
}
static inline void h4_decode(const uint16_t e[4], uint8_t b[4]){
    uint32_t X=e[0],Y=e[1],Z=e[2],W=e[3];
    b[0]=(uint8_t)((X+Y+Z+W-3u*BIAS)>>2);
    b[1]=(uint8_t)(((X+Z+BIAS)-(Y+W))>>2);
    b[2]=(uint8_t)(((X+Y+BIAS)-(Z+W))>>2);
    b[3]=(uint8_t)(((X+W+BIAS)-(Y+Z))>>2);
}
static inline uint32_t h4_l1(const uint16_t a[4], const uint16_t b[4]){
    uint32_t s=0; for(int i=0;i<4;i++){ uint32_t hi=a[i]>b[i]?a[i]:b[i], lo=a[i]>b[i]?b[i]:a[i]; s+=hi-lo;} return s;
}

int main(int argc,char**argv){
    int N = argc>1? atoi(argv[1]) : 13184;
    uint8_t*  B = malloc(N*4);
    uint16_t* E = malloc(N*4*sizeof(uint16_t));
    srand(42);
    for(int i=0;i<N*4;i++) B[i]=(uint8_t)((i*167+13)&0xFF);
    // correctness: round-trip exact + checksum (bit-exactness across languages)
    int errs=0; uint64_t csum=0;
    for(int i=0;i<N;i++){ uint16_t e[4]; uint8_t d[4];
        h4_encode(&B[i*4],e); h4_decode(e,d);
        for(int j=0;j<4;j++){ if(d[j]!=B[i*4+j]) errs++; csum=csum*131+e[j]; E[i*4+j]=e[j]; } }
    // speed: encode timing
    struct timespec t0,t1; clock_gettime(CLOCK_MONOTONIC,&t0);
    volatile uint64_t sink=0;
    for(int r=0;r<1000;r++) for(int i=0;i<N;i++){ uint16_t e[4]; h4_encode(&B[i*4],e); sink+=e[0]; }
    clock_gettime(CLOCK_MONOTONIC,&t1);
    double ns=((t1.tv_sec-t0.tv_sec)*1e9+(t1.tv_nsec-t0.tv_nsec))/(1000.0*N);
    printf("C   : N=%d  roundtrip_errors=%d  checksum=%llu  encode=%.4f ns/box\n",
           N,errs,(unsigned long long)csum,ns);
    free(B);free(E); return errs;
}

File: ports/h4_core.rs
// H4 planogram engine core — Rust. Bit-exact with C/Python (same deterministic input).
const BIAS: u32 = 512;
#[inline] fn h4_encode(b:&[u8;4])->[u16;4]{
    let (x,y,w,h)=(b[0] as u32,b[1] as u32,b[2] as u32,b[3] as u32);
    [(x+y+w+h) as u16, ((x+w+BIAS)-(y+h)) as u16, ((x+y+BIAS)-(w+h)) as u16, ((x+h+BIAS)-(y+w)) as u16]
}
#[inline] fn h4_decode(e:&[u16;4])->[u8;4]{
    let (xx,yy,zz,ww)=(e[0] as u32,e[1] as u32,e[2] as u32,e[3] as u32);
    [((xx+yy+zz+ww-3*BIAS)>>2) as u8, (((xx+zz+BIAS)-(yy+ww))>>2) as u8,
     (((xx+yy+BIAS)-(zz+ww))>>2) as u8, (((xx+ww+BIAS)-(yy+zz))>>2) as u8]
}
fn main(){
    let n:usize = std::env::args().nth(1).and_then(|s|s.parse().ok()).unwrap_or(13184);
    let mut b=vec![0u8;n*4];
    for i in 0..n*4 { b[i]=((i*167+13)&0xFF) as u8; }
    let (mut errs,mut csum)=(0u64,0u64);
    for i in 0..n {
        let bx=[b[i*4],b[i*4+1],b[i*4+2],b[i*4+3]];
        let e=h4_encode(&bx); let d=h4_decode(&e);
        for j in 0..4 { if d[j]!=bx[j]{errs+=1;} csum=csum.wrapping_mul(131).wrapping_add(e[j] as u64); }
    }
    let t0=std::time::Instant::now(); let mut sink=0u64;
    for _ in 0..1000 { for i in 0..n { let bx=[b[i*4],b[i*4+1],b[i*4+2],b[i*4+3]]; sink=sink.wrapping_add(h4_encode(&bx)[0] as u64);} }
    let ns=t0.elapsed().as_nanos() as f64/(1000.0*n as f64);
    println!("Rust: N={}  roundtrip_errors={}  checksum={}  encode={:.4} ns/box  (sink={})",n,errs,csum,ns,sink&1);
}
